movai_core_shared/core/zmq_server.py

```diff
@@ -0,0 +1,166 @@
+"""
+   Copyright (C) Mov.ai  - All Rights Reserved
+   Unauthorized copying of this file, via any medium is strictly prohibited
+   Proprietary and confidential
+
+   Usage:
+        Basic 0MQ client for connecting 0MQ servers.
+
+   Developers:
+   - Erez Zomer (erez@mov.ai) - 2023
+"""
+from abc import ABC, abstractmethod
+import asyncio
+import sys
+import json
+import logging
+import zmq
+import zmq.asyncio
+
+from movai_core_shared.exceptions import UnknownRequestError
+from movai_core_shared.messages.general_data import Request
+
+    
+class ZMQServer(ABC):
+    """
+    This class is a base class for any ZMQ server.
+    """
+    def __init__(self, server_name: str, bind_addr: str, new_loop: bool = False, debug: bool = False) -> None:
+        """Constructor"""
+        if not isinstance(server_name, str):
+            raise ValueError("server name must be of type string.")
+        if not isinstance(bind_addr, str):
+            raise ValueError("bind address must be of type string.")
+        if not isinstance(new_loop, bool):
+            raise ValueError("new_loop must be of type bool.")
+        if not isinstance(debug, bool):
+            raise ValueError("debug must be of type bool.")
+        self._name = server_name
+        self._addr = bind_addr
+        self._logger = logging.getLogger(server_name)
+        self._new_loop = new_loop
+        self._debug = debug
+        self._initialized = False
+        self._running = False
+        self._ctx = None
+        self._socket = None
+
+    def _set_context(self) -> None:
+        """Initializes the zmq context."""
+        self._ctx = zmq.asyncio.Context()
+        self._socket = self._ctx.socket(zmq.ROUTER)
+        if self._socket is None:
+            raise zmq.ZMQError(msg= "Failed to create socket")
+
+    def _bind(self) -> None:
+        """Binds the zmq socket to the appropriate file/address."""
+        self._socket.bind(self._addr)
+        self._logger.info(f"{self.__class__.__name__} is listening on {self._addr}")
+
+    async def _accept(self) -> None:
+        """accepts new connections requests to zmq."""
+        while self._running:
+            try:
+                if self._debug:
+                    self._logger.debug("Waiting for new requests.\n")
+                request = await self._socket.recv_multipart()
```

Comment:
ROUTER type is not thread safe
use with mutex
create lock in __init__
`self._lock = threading.Lock()`
and use it when you send / recieve 
`with self._lock:`

Reply:
unless only one thread will use this logic

Reply:
currently its implemented as single thread with asyncio, do you still think I should add a lock ?

Reply:
lock has been added